sw/device/tests/aon_timer_wakeup_irq_test.c

```diff
@@ -0,0 +1,126 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/freestanding/limits.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_aon_timer.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+static const uint32_t kAonTimerFreq = 200 * 1000;  // 200kHz
+static const uint32_t kSleepTime_us = 2 * 1000;    // 2ms
+static dif_aon_timer_t aon_timer;
+static dif_rv_plic_t plic;
+static dif_rv_plic_irq_id_t plic_irq_id;
+static ibex_timeout_t timeout;
+
+static void aon_timer_wakeup_config(dif_aon_timer_t *aon_timer,
+                                    uint32_t wakeup_threshold) {
+  // Make sure that wake-up timer is stopped.
+  CHECK_DIF_OK(dif_aon_timer_wakeup_stop(aon_timer));
+
+  // Make sure the wake-up IRQ is cleared to avoid false positive.
+  CHECK_DIF_OK(dif_aon_timer_irq_acknowledge(aon_timer,
+                                             kDifAonTimerIrqWkupTimerExpired));
+
+  // Check whether the irq is pending
+  bool is_pending;
+  CHECK_DIF_OK(dif_aon_timer_irq_is_pending(
+      aon_timer, kDifAonTimerIrqWkupTimerExpired, &is_pending));
+  CHECK(!is_pending);
+
+  // Start timer
+  CHECK_DIF_OK(dif_aon_timer_wakeup_start(aon_timer, wakeup_threshold, 0));
+}
+
+static void execute_test(dif_aon_timer_t *aon_timer) {
+  plic_irq_id = kTopEarlgreyPlicPeripheralUnknown;
+
+  LOG_INFO("Wainting wakeup interruption");
+
+  // Capture the current tick to measure the time the irq will take
+  timeout = ibex_timeout_init(kSleepTime_us * 1000);
+
+  // Setup the timer and wait the irq
+  uint32_t sleep_cycles = (kSleepTime_us * kAonTimerFreq / 1000000);
+  aon_timer_wakeup_config(aon_timer, sleep_cycles);
+  wait_for_interrupt();
```

Comment:
i think you probably need to double check that the right interrupt occurred here.

Maybe set a global flag at the end of the ISR or something similar?

Reply:
Looks like the checks are done in the ISR.

Reply:
I think we still do need to do checks after calling `wait_for_interrupt()` as `wait_for_interrupt()` returning does not guarantee that either an interrupt occurred (although Ibex as currently implemented will behave this way) or that the correct interrupt occurred. I think the checks @doreis-lowrisc added should suffice.

One additional point though is that we should probably be calling `wait_for_interrupt()` in a loop until the correct interrupt is seen.